File: mcp/src/velum_mcp/tools/projects.py
"""MCP tools — Projects CRUD."""

from __future__ import annotations

from typing import Any

from velum_mcp.client import get_client


async def list_projects() -> list[dict[str, Any]]:
    """List all Velum projects the current token has access to."""
    client = get_client()
    return await client.get("/projects") or []


async def get_project(project_id: int) -> dict[str, Any]:
    """Get full details of a specific project by ID."""
    client = get_client()
    return await client.get(f"/projects/{project_id}")


async def create_project(
    name: str,
    alert: bool = False,
    alert_chat: str | None = None,
    max_parallel_tasks: int = 0,
) -> dict[str, Any]:
    """Create a new Velum project.

    Args:
        name: Project display name.
        alert: Enable task failure alerts.
        alert_chat: Chat/Telegram ID for alerts.
        max_parallel_tasks: Max tasks running simultaneously (0 = unlimited).
    """
    client = get_client()
    body: dict[str, Any] = {
        "name": name,
        "alert": alert,
        "max_parallel_tasks": max_parallel_tasks,
    }
    if alert_chat:
        body["alert_chat"] = alert_chat
    return await client.post("/projects", body)


async def update_project(
    project_id: int,
    name: str | None = None,
    alert: bool | None = None,
    alert_chat: str | None = None,
    max_parallel_tasks: int | None = None,
) -> dict[str, Any]:
    """Update an existing project's settings."""
    client = get_client()
    current = await client.get(f"/projects/{project_id}")
    body = {
        "name": name if name is not None else current["name"],
        "alert": alert if alert is not None else current.get("alert", False),
        "alert_chat": alert_chat if alert_chat is not None else current.get("alert_chat", ""),
        "max_parallel_tasks": max_parallel_tasks
        if max_parallel_tasks is not None
        else current.get("max_parallel_tasks", 0),
    }
    return await client.put(f"/projects/{project_id}", body)


async def delete_project(project_id: int) -> str:
    """Delete a project and all its resources. This action is irreversible."""
    client = get_client()
    await client.delete(f"/projects/{project_id}")
    return f"Project {project_id} deleted successfully."

File: mcp/src/tools/projects.rs
use crate::client::VelumClient;
use crate::protocol::{prop_bool, prop_int, prop_int_opt, prop_str, prop_str_opt, Tool, ToolResult};
use serde_json::{json, Value};
use anyhow::Result;

pub fn definitions() -> Vec<Tool> {
    vec![
        Tool {
            name: "list_projects".into(),
            description: "List all Velum projects the API token has access to.".into(),
            input_schema: json!({ "type": "object", "properties": {}, "required": [] }),
        },
        Tool {
            name: "get_project".into(),
            description: "Get full details of a specific project.".into(),
            input_schema: json!({
                "type": "object",
                "properties": { "project_id": prop_int("Project ID") },
                "required": ["project_id"]
            }),
        },
        Tool {
            name: "create_project".into(),
            description: "Create a new Velum project.".into(),
            input_schema: json!({
                "type": "object",
                "properties": {
                    "name": prop_str("Project display name"),
                    "alert": prop_bool("Enable task failure alerts"),
                    "alert_chat": prop_str_opt("Chat ID for Telegram alerts"),
                    "max_parallel_tasks": prop_int("Max simultaneous tasks (0 = unlimited)")
                },
                "required": ["name"]
            }),
        },
        Tool {
            name: "update_project".into(),
            description: "Update an existing project's settings.".into(),
            input_schema: json!({
                "type": "object",
                "properties": {
                    "project_id": prop_int("Project ID"),
                    "name": prop_str_opt("New display name"),
                    "alert": { "type": ["boolean","null"], "description": "Toggle alerts" },
                    "max_parallel_tasks": prop_int_opt("Max simultaneous tasks")
                },
                "required": ["project_id"]
            }),
        },
        Tool {
            name: "delete_project".into(),
            description: "Delete a project and all its resources. Irreversible.".into(),
            input_schema: json!({
                "type": "object",
                "properties": { "project_id": prop_int("Project ID to delete") },
                "required": ["project_id"]
            }),
        },
    ]
}

pub async fn call(name: &str, args: &Value, client: &VelumClient) -> Result<ToolResult> {
    match name {
        "list_projects" => {
            let v = client.get("/projects").await?;
            Ok(ToolResult::json(&v))
        }
        "get_project" => {
            let id = args["project_id"].as_i64().unwrap_or_default();
            let v = client.get(&format!("/projects/{id}")).await?;
            Ok(ToolResult::json(&v))
        }
        "create_project" => {
            let body = json!({
                "name": args["name"],
                "alert": args.get("alert").and_then(Value::as_bool).unwrap_or(false),
                "alert_chat": args.get("alert_chat").and_then(Value::as_str).unwrap_or(""),
                "max_parallel_tasks": args.get("max_parallel_tasks").and_then(Value::as_i64).unwrap_or(0)
            });
            let v = client.post("/projects", &body).await?;
            Ok(ToolResult::json(&v))
        }
        "update_project" => {
            let id = args["project_id"].as_i64().unwrap_or_default();
            let mut current = client.get(&format!("/projects/{id}")).await?;
            if let Some(name) = args.get("name").and_then(Value::as_str) {
                current["name"] = json!(name);
            }
            if let Some(alert) = args.get("alert").and_then(Value::as_bool) {
                current["alert"] = json!(alert);
            }
            if let Some(mpt) = args.get("max_parallel_tasks").and_then(Value::as_i64) {
                current["max_parallel_tasks"] = json!(mpt);
            }
            let v = client.put(&format!("/projects/{id}"), &current).await?;
            Ok(ToolResult::json(&v))
        }
        "delete_project" => {
            let id = args["project_id"].as_i64().unwrap_or_default();
            client.delete(&format!("/projects/{id}")).await?;
            Ok(ToolResult::text(format!("Project {id} deleted successfully.")))
        }
        _ => Ok(ToolResult::error(format!("Unknown tool: {name}"))),
    }
}
